framework/src/main/java/org/checkerframework/framework/flow/CFAbstractStore.java

```diff
@@ -249,39 +249,73 @@ public void updateForMethodCall(
             || (assumePureGetters && ElementUtils.isGetter(method))
             || atypeFactory.isSideEffectFree(method));
     if (hasSideEffect) {
+      localVariableValues
+          .values()
+          .forEach(
+              v -> {
+                if (v != null && v.getThenStore() != null) {
+                  v.getThenStore().updateForSideEffect(gatypeFactory);
+                  v.getElseStore().updateForSideEffect(gatypeFactory);
+                }
+              });
+      updateForSideEffect(gatypeFactory);
+    }
 
-      boolean sideEffectsUnrefineAliases = gatypeFactory.sideEffectsUnrefineAliases;
-
-      // update local variables
-      // TODO: Also remove if any element/argument to the annotation is not
-      // isUnmodifiableByOtherCode.  Example: @KeyFor("valueThatCanBeMutated").
-      if (sideEffectsUnrefineAliases) {
-        localVariableValues.entrySet().removeIf(e -> e.getKey().isModifiableByOtherCode());
-      }
-
-      // update this value
-      if (sideEffectsUnrefineAliases) {
-        thisValue = null;
-      }
+    // store information about method call if possible
+    JavaExpression methodCall = JavaExpression.fromNode(methodInvocationNode);
+    replaceValue(methodCall, val);
+  }
 
-      // update field values
-      if (sideEffectsUnrefineAliases) {
-        fieldValues.entrySet().removeIf(e -> e.getKey().isModifiableByOtherCode());
-      } else {
-        // Case 2 (unassignable fields) and case 3 (monotonic fields)
-        updateFieldValuesForMethodCall(gatypeFactory);
-      }
+  /**
+   * Remove any information that might not be valid any more after a method call, and add
+   * information guaranteed by the method.
+   *
+   * <ol>
+   *   <li>If the method is side-effect-free (as indicated by {@link
+   *       org.checkerframework.dataflow.qual.SideEffectFree} or {@link
+   *       org.checkerframework.dataflow.qual.Pure}), then no information needs to be removed.
+   *   <li>Otherwise, all information about field accesses {@code a.f} needs to be removed, except
+   *       if the method {@code n} cannot modify {@code a.f}. This unmodifiability property holds if
+   *       {@code a} is a local variable or {@code this}, and {@code f} is final, or if {@code a.f}
+   *       has a {@link MonotonicQualifier} in the current store. Subclasses can change this
+   *       behavior by overriding {@link #newFieldValueAfterMethodCall(FieldAccess,
+   *       GenericAnnotatedTypeFactory, CFAbstractValue)}.
+   *   <li>Furthermore, if the field has a monotonic annotation, then its information can also be
```

Comment:
What is "the field"?

Reply:
What is the relationship between "has a monotonic annotation" here and "has a `MonotonicQualifier`" four lines above?